### DISCUSSION TITLE
forecasts with WindowSummarizer in a pipeline should remember past data (but currently don't)

### DISCUSSION THREAD

**Discussion by GrzegorzRut** [alan-turing-institute/sktime#2309] [Q&A]
Hi!
I'm looking for a function that would allow to create a window summary of the data points from the past (endog or exog) that one could use for forecasting. I thought that `WindowSummarizer` is designed just for that but maybe it's expected to be used in different scenarios.

There is a question I posted a question on [stack overflow](https://stackoverflow.com/questions/71561875/forecasting-with-windowsummarizer-and-exogenous-features) regarding the usage of WindowSummarizer for forecasting. The [documentation](https://www.sktime.org/en/stable/api_reference/auto_generated/sktime.transformations.series.window_summarizer.WindowSummarizer.html?highlight=WindowSummarizer#windowsummarizer) uses a `NaiveForecaster` which, as far as I know, will not be able to take advantage from new features. Replacing it with an algorithm that allows for exogenous features leads to errors.

I think that this is the same issue as the one that came up in the [discussion](https://github.com/alan-turing-institute/sktime/discussions/2267) on the `Imputer`. The window functions do not remember values from training data, they are recalculated from scratch for unseen data. This means that there is a disconnection between the training data and the new data points.

Here's a minimal working example (focused on a time series without exog features for simplicity):

```
from sktime.forecasting.base import ForecastingHorizon
from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveForecaster
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.compose import ForecastingPipeline
from sktime.transformations.series.window_summarizer import WindowSummarizer

import pandas as pd
y = load_airline()
y_train, y_test = temporal_train_test_split(y.iloc[:10])
fh = ForecastingHorizon(y_test.index, is_relative=False)

kwargs = {
    "lag_config": {
        "mean": ["mean", [[3, 1], [4, 1]]],
    }
}

pipe = ForecastingPipeline(
    steps=[
        ("ws", WindowSummarizer(**kwargs, n_jobs=1)),
        ("forecaster", NaiveForecaster(strategy='drift')),
    ]
)
pipe_return = pipe.fit(y_train,pd.DataFrame(y_train))
y_pred = pipe_return.predict(fh=fh,X=pd.DataFrame(y_test)) 
ws = pipe_return.steps_[0][1]
ws.transform(pd.DataFrame(y_test))
```
<div class="lm-Widget p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output" data-mime-type="text/html"><div>

  | Number of airline passengers_mean_3_1 | Number of airline passengers_mean_4_1
-- | -- | --
1949-08 | NaN | NaN
1949-09 | NaN | NaN
1949-10 | NaN | NaN

</div></div>
A work around would be to save during fitting a minimal subset of the training data required for a continuity between the past and the unseen (for instance, as the instance's attribute). This would get a bit complicated with updates but I have no better idea how to approach this issue.

Is there a different function that is better for these kinds of tasks?

**Comment by GrzegorzRut** [comment]
Hi!
I was able to figure out a work around. I've posted the answer [here](https://stackoverflow.com/questions/71561875/forecasting-with-windowsummarizer-and-exogenous-features).

There are some tricky parts - I had to make a choice between keeping a minimal window required for forecasting or saving all seen X records (the second approach seemed a bit more natural).

The code is quite crude and I'm not sure I was able to catch all exceptions. I'd welcome all suggestions how to improve it.

**Reply by fkiraly** [reply]
Thanks!
So, I did not take your code directly, but I took your idea in #2325.

Questions:
* I added you as an author. Is that ok?
* if you would like, you could add yourself to `all-contributorsrc` with an `"ideas"` contribution, this is already the second time you point out sth very useful. Kindly PR yourself if you want.
* would you like to contribute to the code base directly? Feel free to make a PR! You could base yours on `main` or my branch.

**Reply by GrzegorzRut** [reply]
> I added you as an author. Is that ok?

Sure, thank you.

> if you would like, you could add yourself to all-contributorsrc with an "ideas" contribution, this is already the second time you point out > sth very useful. Kindly PR yourself if you want.

Sure, I'd be happy to join.

> would you like to contribute to the code base directly? Feel free to make a PR! You could base yours on main or my branch.

I'm new to open source and collaboration in general but I'd like to give it a try. Thank you!

**Comment by fkiraly** [comment]
> The window functions do not remember values from training data, they are recalculated from scratch for unseen data.

Yes, I think you are right, @GrzegorzRut.
Same problem as with the imputer.

Imo, the `WindowSummarizer` should be doing what your version of it does!
It should not be "remembering" the transformed data on the training set, and "playing it back" on the test/prediction set, instead of calculating them in `transform`!

FYI @danbartl

**Comment by fkiraly** [comment]
@GrzegorzRut, I've tried out a fix here:
https://github.com/alan-turing-institute/sktime/pull/2325

Approach is giving the `WindowSummarizer` memory, concatenating remembered `X` in `transform`, then subsetting to the indices seen in `transform`. Bit wasteful as we could cut off any part of the `fit`-`X` that do not participate in a summary window ending up in the `transform` indices, but it's a start to see whether this is what we need.

**Reply by fkiraly** [reply]
yes, seems to work.

Here's what your code example above now returns, on the #2325 branch:

<html>
<body>


. | Number of airline passengers_mean_3_1 | Number of airline passengers_mean_4_1
-- | -- | --
1949-08 | 128.333333 | 129.25
1949-09 | 134.666667 | 133.25
1949-10 | 143.666667 | 138.00


</body>
</html>

**Reply by GrzegorzRut** [reply]
Right, I'm aware that this is not efficient but the most efficient way (commented out in the code in `update_X`)
`# self.X_window = X_window.iloc[-self.min_window:]` 
also has drawbacks.

The problem with this approach is that it works well only if we apply `transform` only once on a particular set of data. This might not be an issue but I guess that getting a different result after the second `transform` (the first one will modify the saved window) would be considered awkward.

To some degree, this 'wastefulness' could be fixed by changing the `_fit` method but this might be considered a half measure.
```
    def _fit(self, X, y=None):
        update_X(self,X.iloc[-self.min_window:]) # specifying which training records should be saved
        super()._fit(X, y)
```

**Reply by fkiraly** [reply]
> The problem with this approach is that it works well only if we apply

I don't see this - `transform` would not modify/update the memory. Only `_fit` and `_update` should do that.
Can you explain, please?

I thought your approach makes sense, since no invocation of `transform` will ever ask for data that is older than `-min_window`.

**Reply by fkiraly** [reply]
> To some degree, this 'wastefulness' could be fixed

I don't think you should inherit from `WindowSummarizer`, we just need to fix `WindowSummarizer` itself to make it work properly.
As usualy in an open source project, much of the codebase is work in progress, and any contributions are appreciated!
😄

**Reply by GrzegorzRut** [reply]
> The problem with this approach is that it works well only if we apply
This problem **would** appear (I was referring to the approach that is commented out in my post on stack overflow), if we'd use
```
def update_X(self,X):
    if self.target_cols==None:
        cols = X.columns
    else:
        cols = self.target_cols
    X_window = self.X_window
    X_window = pd.concat([X_window,X[cols]],axis=0)
    X_window = X_window.groupby(X_window.index).first()
    # would remember only last #min_window rows
    self.X_window = X_window.iloc[-self.min_window:]
```
instead of

```
def update_X(self,X):
    if self.target_cols==None:
        cols = X.columns
    else:
        cols = self.target_cols
    X_window = self.X_window
    X_window = pd.concat([X_window,X[cols]],axis=0)
    X_window = X_window.groupby(X_window.index).first()
    # would remember all rows
    self.X_window = X_window
```

> I don't see this - transform would not modify/update the memory. Only _fit and _update should do that.
> Can you explain, please?

In my code this update is done as follows:
```
    def _transform(self, X, y=None):
        X_window = pd.concat([self.X_window,X],axis=0)
        X_window = X_window.groupby(X_window.index).first()
        X_transformed = super()._transform(X_window, y)
        update_X(self,X) # <-- **this is the update**
        return X_transformed.loc[X.index]
```
I decided to do the update after I'd read the code in `ForecastingPipeline`  for `_predict` . It calls `transform`, so it felt natural to update the window in `transform` for everything to work. Not the prettiest code..

**Reply by fkiraly** [reply]
ah, I see - the problem is probably resolved if I explain how the `sktime` transformer interface is meant to work.

`transform` should never write to `self`, it is a non-state-changing method.

If you want to use transformers in a stream/forecasting setting, you should call the following methods in sequence:
* `fit` on the first chunk
* `update` on the second chunk
* `transform` on the second chunk
* `update` on the third chunk
* `transform` on the third chunk

If you use it in a `ForecastingPipeline`, the following is applied, in sequence:
* `transformer.fit` on the first chunk `X`
* `forecaster.fit` on the first chunk `y` and transformed first chunk `X`
* `forecaster.predict` from the first chunk, on transformed first chunk `X`
* `transformer.update` on the second chunk `X`
* `transformer.transform` on the second chunk `X`
* `forecaster.update` on the second chunk, on transformed second chunk `X`
* `forecaster.predict` from the second chunk, on transformed second chunk `X`
* and so on

Hence, no need to update in `_transform`, just do it in `_update`, which is expressly meant for that!

**Reply by fkiraly** [reply]
(you seem to assume that the sequence for transformers is `fit`, `transform`, `transform`, etc, which - as you say - would cause a problem with multiple application of `transform`)

**Comment by fkiraly** [comment]
actually, separate topic - I think your code, could be a nice basis for a "transform together with fit" wrapper!

I.e., remember the data from fit/update, transform all of it together, and then subset. This can in general have a result different from "just use the data on transform".

The code you wrote could, in-principle, wrap around any transformer (if we turn it into a wrapper).

**Reply by GrzegorzRut** [reply]
I'd be happy to help!

**Reply by fkiraly** [reply]
would appreciate a sanity check here, i.e., that this is not already possible with some interface element more easily.
E.g., first calling `fit` on some `X` and then calling `transform` on a larger `X`.

(I think not, but it wouldn't be the first time I get confused)

**Reply by GrzegorzRut** [reply]
1. `pd.concat([self._X_memory, X]) ` - this assumes that indices in X are not the same as in `self._X_memory`.
Should the code try to deduplicate records by default or raise a flag that some index values are duplicated?

2. I'm not sure if this is a problem, should data be sorted on the index? This is connected with the 'issue' above. Both points are a bit artificial, I guess.

3. Should there be a check on the completeness of an index?
For instance, if `self._X_memory` has [1940,1941] in the index and X has  [1943,1944] - should the updated memory have
[1940,1941,1942, 1943, 1944] or throw an error?

4. In principle we can focus only on `self._target_cols` in the memory - the rest is not relevant for the window functions.
This would require to have 
`self._X_memory = X[self._target_cols]` 
after the definition of `self._target_cols` .

5. I was not able to confirm that the code integrates well with other classes like `ForecastingPipeline`. Wrote about it on slack.

**Reply by fkiraly** [reply]
Thanks for the review! I'll reply in the PR #2325, following general good practice that discussion about the code should be, if possible, attached to the code (in this case the PR).

**Reply by GrzegorzRut** [reply]
Ok, thank you!

**Comment by fkiraly** [comment]
so, @GrzegorzRut, quick update:

0.11.3 was recently released, and `WindowSummarizer` should now have memory, as discussed above!
See https://github.com/alan-turing-institute/sktime/pull/2325

The imputer, which was also mentioned, is in progress here and should be ready for 0.12.0:
https://github.com/alan-turing-institute/sktime/pull/2382

**Reply by GrzegorzRut** [reply]
Many thanks!
